File: lab2/JS/Singleton.js
let singleton = Symbol();
let singletonEnforcer = Symbol();

class Singleton {
    constructor(enforcer) {
        this.LinksList = {}
        if (enforcer !== singletonEnforcer)
            throw "Instantiation failed: use Singleton.getInstance() instead of new.";
    }
    static get _instance() {
        if (!this[singleton])
            this[singleton] = new Singleton(singletonEnforcer);
        return this[singleton]
    }
    static set _instance(v){throw "Can't change constant property!"}
    static getInstance(){return this._instance;}
    
    addLink(link){
        if(this.LinksList.hasOwnProperty(link)){
            this.LinksList[link] +=1;
        }
        else{
            this.LinksList[link] = 1;
        }
    }

    printAllLinks(){
        if (Object.keys(this.LinksList).length){
            for (let link in this.LinksList)
                console.log(`Link: ${link} (links added: ${this.LinksList[link]});`);
        }
        else console.log("No links");
    }

    clearAllLinks(){
        this.LinksList = {}
        console.log("Links' list cleared!");
    }
}

export default Singleton

File: lab2/Cs/Singleton.cs
using System;
using System.Collections.Generic;
namespace Singleton
{
    public class Singleton
    {
        private static Singleton _instance;
        private Dictionary<string,int> LinksList { get; set; }

        public static Singleton GetInstance()
        {
            return _instance ?? (_instance = new Singleton());
        }

        private Singleton()
        {
            LinksList = new Dictionary<string, int>();
        }
        public void AddLink(string link)
        {
            if (LinksList.ContainsKey(link))
                LinksList[link] += 1;
            else
                LinksList.Add(link,1);
        }

        public void PrintAllLinks()
        {
            foreach (var link in LinksList)
                Console.WriteLine(link);
        }
        public void ClearAllLinks()
        {
            LinksList.Clear();
            Console.WriteLine("Links' list cleared!");
        }
    }
}